src/OrleansRuntime/MultiClusterNetwork/MultiClusterOracle.cs

```diff
@@ -0,0 +1,766 @@
+            if (localData.Current.Configuration != null)
+            {
+                allClusters = allClusters.Union(localData.Current.Configuration.Clusters);
+            }
+
+            return new HashSet<string>(allClusters);
+        }
+
+        private void RemoveStaleTaskStatusEntries<K>(Dictionary<K, GossipStatus> dict)
+        {
+            var now = DateTime.UtcNow;
+            var toRemove = dict
+                .Where(kvp => (now - kvp.Value.LastUse).TotalMilliseconds > 2.5 * this.randomizedResendActiveStatusAfter.TotalMilliseconds)
+                .Select(kvp => kvp.Key)
+                .ToList();
+
+            foreach (var key in toRemove)
+                dict.Remove(key);
+        }
+      
+        // called by remote nodes that push changes
+        public Task Push(IMultiClusterGossipData gossipData, bool forwardLocally)
+        {
+            logger.Verbose("--- Push: receive {0} data {1}", forwardLocally ? "remote" : "local", gossipData);
+
+            var data = (MultiClusterData)gossipData;
+
+            var delta = localData.ApplyDataAndNotify(data);
+
+            // forward changes to all local silos
+            if (forwardLocally)
+            {
+                foreach (var activeSilo in this.GetApproximateOtherActiveSilos())
+                    PushGossipToSilo(activeSilo, delta);
+            }
+
+            PushMyStatusToNewDestinations(delta);
+
+            logger.Verbose("--- Push: done");
+
+            return TaskDone.Done;
+        }
+
+        // called by remote nodes' full background gossip
+        public Task<IMultiClusterGossipData> PushAndPull(IMultiClusterGossipData gossipData)
+        {
+            logger.Verbose("--- PushAndPull: gossip {0}", gossipData);
+
+            var data = (MultiClusterData)gossipData;
+
+            var delta = this.localData.ApplyDataAndNotify(data);
+
+            PushMyStatusToNewDestinations(delta);
+
+            logger.Verbose("--- PushAndPull: done, answer={0}", delta);
+
+            return Task.FromResult((IMultiClusterGossipData)delta);
+        }
+
+        public async Task<Dictionary<SiloAddress, MultiClusterConfiguration>> CheckMultiClusterStability(MultiClusterConfiguration expected)
+        {
+            var tasks = new List<Task<Dictionary<SiloAddress, MultiClusterConfiguration>>>();
+            tasks.Add(FindUnstableSilos(expected, true));
+
+            foreach (var cluster in GetActiveClusters())
+            {
+                if (cluster != this.clusterId)
+                {
+                    var silo = GetRandomClusterGateway(cluster);
+                    if (silo == null)
+                        throw new OrleansException("no gateway for cluster " + cluster);
+                    var remoteOracle = InsideRuntimeClient.Current.InternalGrainFactory.GetSystemTarget<IMultiClusterGossipService>(Constants.MultiClusterOracleId, silo);
+                    tasks.Add(remoteOracle.FindUnstableSilos(expected, true));
```

Comment:
By design the multicluster network must be able to tolerate network partitions, i.e. there should not be any exceptions propagating up and breaking things.

Currently, we rely on the background gossip to re-synchronize data after partition heals. We should probably measure the speed of that though - perhaps we need some optimization.